src/OpenTelemetry.AutoInstrumentation/Loading/AssemblyLoadDetector.cs

```diff
@@ -0,0 +1,67 @@
+// <copyright file="AssemblyLoadDetector.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+
+namespace OpenTelemetry.AutoInstrumentation.Loading
+{
+    internal abstract class AssemblyLoadDetector
+    {
+        private HashSet<string> _requiredAssemblies;
+        private int _loadedCount;
+
+        public AssemblyLoadDetector(IEnumerable<string> requiredAssemblies)
+        {
+            _requiredAssemblies = new HashSet<string>(requiredAssemblies);
+
+            AppDomain.CurrentDomain.AssemblyLoad += CurrentDomain_AssemblyLoad;
+        }
+
+        public event EventHandler<LoadDetectorReadyEventArgs> OnReady;
+
+        internal abstract Func<object> GetInstrumentationBuilder();
+
+        private void CurrentDomain_AssemblyLoad(object sender, AssemblyLoadEventArgs args)
+        {
+            string assemblyName = args.LoadedAssembly.FullName.Split(',')[0];
+
+            if (_requiredAssemblies.Contains(assemblyName, StringComparer.InvariantCultureIgnoreCase))
+            {
+                if (Interlocked.Increment(ref _loadedCount) == _requiredAssemblies.Count)
```

Comment:
Can you imagine any scenarios where multiple versions of the same assembly name are loaded? That could mess up the comparison with `loadedCount`

Reply:
That's good point. NET Framework should be safe until we watch only default appdomain, in .NET Core I think I need to fix it a bit, by looking only into default assembly load context.

Reply:
That's already a way bigger issue, we are not generating spans in sibling AppDomains / load contextes as the SDK is loaded into default context.

Reply:
Then I suggest only watching the default AppDomain/LoadContext. It would be good to mention it in `troubleshooting.md` that auto-instrumentation is only working for the default AppDomain/LoadContext.

Reply:
I would extract an issue from this comment. Need to build some tests apps to verify which parts do work and which not. With source based instrumentation I have some idea and it's already partially tested, but I'm not sure how rejited libraries are selected - it probably doesn't not look for appdomain / load context.